rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -341,72 +363,233 @@ def _convert_attribute_tokens_to_texts(
-
-        self.vectorizers = self._create_shared_vocab_vectorizers(
-            {
-                "strip_accents": self.strip_accents,
-                "lowercase": self.lowercase,
-                "stop_words": self.stop_words,
-                "min_ngram": self.min_ngram,
-                "max_ngram": self.max_ngram,
-                "max_df": self.max_df,
-                "min_df": self.min_df,
-                "max_features": self.max_features,
-                "analyzer": self.analyzer,
-            }
-        )
+    @staticmethod
+    def _get_starting_empty_index(vocabulary: Dict[Text, int]) -> int:
+        for key in vocabulary.keys():
+            if key.startswith(BUFFER_SLOTS_PREFIX):
+                return int(key.split(BUFFER_SLOTS_PREFIX)[1])
+        return len(vocabulary)
+
+    def _update_vectorizer_vocabulary(
+        self, attribute: Text, vocabulary: Set[Text]
+    ) -> None:
+        """Update the existing vocabulary of the vectorizer with new unseen words.
 
-        combined_cleaned_texts = []
-        for attribute in self._attributes:
-            combined_cleaned_texts += attribute_texts[attribute]
+        These unseen words should only occupy the empty buffer slots.
 
-        try:
-            self.vectorizers[TEXT].fit(combined_cleaned_texts)
-        except ValueError:
+        Args:
+            attribute: Message attribute for which vocabulary should be updated.
+            vocabulary: Set of words to expand the vocabulary with if they are unseen.
+
+        """
+        existing_vocabulary: Dict[Text, int] = self.vectorizers[attribute].vocabulary
+        available_empty_index = self._get_starting_empty_index(existing_vocabulary)
+        if len(vocabulary) > len(existing_vocabulary):
             logger.warning(
-                "Unable to train a shared CountVectorizer. "
-                "Leaving an untrained CountVectorizer"
+                f"New data contains vocabulary of size {len(vocabulary)} for attribute {attribute} "
+                f"which is larger than the maximum vocabulary size({len(existing_vocabulary)}) "
+                f"of the original model. Some tokens will have to be dropped "
+                f"in order to continue training. It is advised to re-train the "
+                f"model from scratch on complete data."
             )
+        for token in vocabulary:
+            if token not in existing_vocabulary:
+                existing_vocabulary[token] = available_empty_index
+                del existing_vocabulary[f"{BUFFER_SLOTS_PREFIX}{available_empty_index}"]
+                available_empty_index += 1
+                if available_empty_index == len(existing_vocabulary):
+                    break
+        self._set_vocabulary(attribute, existing_vocabulary)
+
+    def _get_additional_vocabulary_size(
+        self, attribute: Text, existing_vocabulary_size: int
+    ) -> int:
+        """Get additional vocabulary size to be saved for incremental training.
```

Comment:
We decided to use descriptive docstrings a couple of months ago (also part of the Python code conventions)
```suggestion
        """Gets additional vocabulary size to be saved for incremental training.
```